database-commons/src/main/java/io/cdap/plugin/db/connector/AbstractDBSpecificConnector.java

```diff
@@ -0,0 +1,124 @@
+
+  protected final AbstractDBSpecificConnectorConfig config;
+
+  protected AbstractDBSpecificConnector(AbstractDBSpecificConnectorConfig config) {
+    super(config);
+    this.config = config;
+  }
+
+  public abstract boolean supportSchema();
+
+  protected abstract Class<? extends DBWritable> getDBRecordType();
+
+  protected SchemaReader getSchemaReader() {
+    return new CommonSchemaReader();
+  }
+
+  @Override
+  protected DBConnectorPath getDBConnectorPath(String path) throws IOException {
+    return DBSpecificPath.of(path, supportSchema());
+  }
+
+  @Override
+  public InputFormatProvider getInputFormatProvider(ConnectorContext context, SampleRequest request)
+    throws IOException {
+    DBSpecificPath path = DBSpecificPath.of(request.getPath(), supportSchema());
+    if (path.getTable() == null) {
+      throw new IllegalArgumentException(
+        String.format("Path %s cannot be sampled. Must have table name in the path.", request.getPath()));
+    }
+
+    ConnectionConfigAccessor connectionConfigAccessor = new ConnectionConfigAccessor();
+    if (config.getUser() == null && config.getPassword() == null) {
+      DBConfiguration
+        .configureDB(connectionConfigAccessor.getConfiguration(), driverClass.getName(),
+                     getConnectionString(path.getDatabase()));
+    } else {
+      DBConfiguration
+        .configureDB(connectionConfigAccessor.getConfiguration(), driverClass.getName(),
+                     getConnectionString(path.getDatabase()),
+                     config.getPassword(), config.getPassword());
+    }
+    String tableQuery = getTableQuery(path);
+    DataDrivenETLDBInputFormat.setInput(connectionConfigAccessor.getConfiguration(), getDBRecordType(),
+                                        tableQuery, null, false);
+    connectionConfigAccessor.setConnectionArguments(Maps.fromProperties(config.getConnectionArgumentsProperties()));
+    connectionConfigAccessor.getConfiguration().setInt(MRJobConfig.NUM_MAPS, 1);
+    try {
+      connectionConfigAccessor.setSchema(loadTableSchema(getConnection(path),  tableQuery).toString());
+    } catch (SQLException e) {
+      throw new IOException(String.format("Failed to get table schema due to: %s.", e.getMessage()), e);
+    }
+
+
+    return new SourceInputFormatProvider(DataDrivenETLDBInputFormat.class, connectionConfigAccessor.getConfiguration());
+  }
+
+  protected Connection getConnection(DBConnectorPath path) {
+    return getConnection(getConnectionString(path.getDatabase()) , config.getConnectionArgumentsProperties());
+  }
+
+  protected String getConnectionString(String database) {
+    return config.getConnectionString();
+  }
+
+  protected String getTableQuery(DBSpecificPath path) {
+    return path.getSchema() == null ? String.format("SELECT * from %s.%s", path.getDatabase(), path.getTable()) : String
+      .format("SELECT * from %s.%s.%s", path.getDatabase(), path.getSchema(), path.getTable());
```

Comment:
Are you sure we need to include database? Cross-datadase queries is not a standard feature. E.g. postgresql docs says "The name (optionally schema-qualified) of an existing table or view". (https://www.postgresql.org/docs/current/sql-select.html). It does not say you can include a database name

Reply:
Also it may produce problems if connection is going through some kind of pooler (e.g. PGPOOL) as there going to be a mapping between pooler DBs and database server DBs

Reply:
Those DBs that don't support Cross-database queries will only connect to the DB being queried. That's why PostgresConnector overrides the `getConnection(database)` methods.

Thanks for bring up the issue for pooler, but could you educate me what the problem will be if the database supports Cross-database queries with pooler ?

Reply:
Well, e.g. in pgbouncer there is next setting that allows to map database names. This means that database in URI and database within remote server are different beasts, see https://www.pgbouncer.org/config.html
This allows a single pooler to pool multiple servers with potentially overlapping database names

Reply:
Discussed offline and will leave the current code as is and will keep in mind this situation.

Reply:
Can you open a JIRA to track and put a TODO here

Reply:
Discussed offline, no need to Add TODO here, filed PLUGIN-714 for documenting that we don't support pooler that proxies multiple db server. Because when we connect to DB initially , we don't specify database name in the connection string, if a pooler is proxying multiple DB server, it won't know which server to connect to.